### PR TITLE
Add blog post for "Building a healthcare chatbot with Mixtral 8x7b, Haystack, and PubMed"

### PR DESCRIPTION
**Pull request by annthurium**

### CODE REVIEW

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+authors:
+  - Tilde Thurium
+tags: ["RAG", "LLM", "Haystack 2.0"]
+---
+
+Unfortunately, countless people around the world have inadequate access to healthcare. I’m lucky to have health insurance and good medical providers taking care of me. However, I still want to educate myself before walking into a doctor’s office.
+
+Technology can empower people to take charge of their health. Large language models can power chatbots where people can ask medical questions. 
+
+In this post, I’ll show you how I built a medical chatbot with Haystack 2.0, and the Mixtral 8x7B model that pulls research papers from PubMed.
```
Comment (TuanaCelik) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:20]:
```suggestion
In this post, I’ll show you how I built a medical chatbot with Haystack 2.0-Beta, and the Mixtral 8x7B model by pulling research papers from PubMed.
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+Building a medical chatbot presents some challenges.
+- *Lack of data*. HIPAA and other privacy regulations make it harder to find public QA datasets to fine-tune a model.
+- The human impact of providing wrong answers. 😬
+- *Staying up to date*. It takes a long time to train a LLM. By the time they are released, their knowledge is already a bit stale. Medical research breakthroughs are happening all the time. I am particularly interested in long COVID, which has both impacted humanity on a massive scale, and is the subject of ongoing research.
+
+Therefore, I decided to use a RAG pipeline to combine PubMed data with a LLM.
+
+## High Level Architecture Overview
+
+Retrieval augmented generative AI, or RAG, is a way of giving a LLM context so it can better answer questions.
```
Comment (TuanaCelik) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:35]:
```suggestion
Retrieval augmented generation, or RAG, is a way of giving a LLM context so it can better answer questions. 
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        documents = [documentize(article) for article in response]
+        articles.extend(documents)
+    except Exception as e:
+        print(e)
+        print(f"Couldn't fetch articles for queries: {queries}" )
+    results = {'articles': articles}
+    return results
+```
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique new kind of model that uses 8 “experts” that queries can be routed to. Each expert has 7 billion parameters, yet the queries can be pretrained quickly and have faster inference.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail.
```
Comment (TuanaCelik) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:74]: I would rephrase this. Because in reality, it's a bit more complex than 'queries' being routed. It's actually a per token routing (as per my understanding).

Instead maybe something like. Mixrtal is a unique kind of model that uses 8 "experts" and an internal "routing" mechanism that routes a token to a specific expert. This also means that during inferencing, not all parameters are used, which allows the model to response remarkably fast.

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        print(e)
+        print(f"Couldn't fetch articles for queries: {queries}" )
+    results = {'articles': articles}
+    return results
+```
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique new kind of model that uses 8 “experts” that queries can be routed to. Each expert has 7 billion parameters, yet the queries can be pretrained quickly and have faster inference.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail. 
+
+
+## Prompts and queries
```
Comment (TuanaCelik) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:77]: To me this could be 2 sections:
1) Generating Keywords for PubMed with Mixtral/LLMs
2) The PubMed RAG pipeline

And you could split the 2 bits of code into these 2 sections too

**Review by TuanaCelik — CHANGES_REQUESTED** [review]

**Comment by annthurium** [comment]
Thank you so much for your feedback @TuanaCelik! And for the lovely header image! I believe I've addressed all the issues. Want to take another look next week after you return?

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+description: Have a look into our first commitment to what will eventually become Haystack 2.0
+featured_image: thumbnail.png
+images: ["blog/mixtral-8x7b-healthcare-chatbot/thumbnail.png"]
+alt_image: The Mixtral logo and the Haystack logo against a white background.
+toc: True
+date: 2024-01-02
+last_updated:  2024-01-02
+authors:
+  - Tilde Thurium
+  - Tuana Celik
```
Comment (TuanaCelik) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:13]: Ah definitely remove me :) - I'm just reviewing, your work!

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+## The PubMed RAG Pipeline
+
+Retrieval augmented generation, or RAG, is a way of giving a LLM context so it can better answer questions. 
+
+You pass the LLM some documents, along with a query, and prompt the LLM to use the documents when answering the question.
+
+![](RAG.png)
+
+[PubMed](https://pubmed.ncbi.nlm.nih.gov/) has up to date, trustworthy medical information so it seemed like a solid document source. Plus, there's a [PyMed wrapper for the PubMed API](https://github.com/gijswobben/pymed) that made querying easy peasy. We'll wrap this in a Haystack custom component to format the results as `Document`s so that Haystack can use them, and add some light error handling.
```
Comment (TuanaCelik) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:42]:
```suggestion
[PubMed](https://pubmed.ncbi.nlm.nih.gov/) has up to date, trustworthy medical information so it seemed like a solid document source. Plus, there's a [PyMed wrapper for the PubMed API](https://github.com/gijswobben/pymed) that made querying easy peasy. We'll wrap this in a [Haystack custom component](https://docs.haystack.deepset.ai/v2.0/docs/custom-components) to format the results as `Document`s so that Haystack can use them, and add some light error handling.
```

**Comment by TuanaCelik** [comment]
Looks good 🎉

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique kind of model that uses 8 *"experts"* and an internal *"routing"* mechanism that routes a token to a specific expert. This also means that during inferencing, not all parameters are used, which allows the model to response remarkably fast.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail. 
+
+
+## Generating Keywords for PubMed with Mixtral/LLMs
+
+First, I tried an approach where I passed a plain query to PubMed. e.g. *"What are the most  current treatments for long COVID?"* Unfortunately, that didn't work too well. The articles returned weren't very relevant. Which makes sense, because PubMed isnt optimized for natural language search. It is optimized for keywords, though. And you know what's great at generating keywords? LLMs!
+
+So now our flow is as follows:
+- Our user inputs a question, such as, "
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:82]: This line seems missing

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+- We prompt the LLM to turn tha question into keywords
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:83]:
```suggestion
- We prompt the LLM to turn the question into keywords
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    results = {'articles': articles}
+    return results
+```
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique kind of model that uses 8 *"experts"* and an internal *"routing"* mechanism that routes a token to a specific expert. This also means that during inferencing, not all parameters are used, which allows the model to response remarkably fast.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail. 
+
+
+## Generating Keywords for PubMed with Mixtral/LLMs
+
+First, I tried an approach where I passed a plain query to PubMed. e.g. *"What are the most  current treatments for long COVID?"* Unfortunately, that didn't work too well. The articles returned weren't very relevant. Which makes sense, because PubMed isnt optimized for natural language search. It is optimized for keywords, though. And you know what's great at generating keywords? LLMs!
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:79]: Loved this section, just a small typo here
```suggestion
First, I tried an approach where I passed a plain query to PubMed. e.g. *"What are the most  current treatments for long COVID?"* Unfortunately, that didn't work too well. The articles returned weren't very relevant. Which makes sense, because PubMed isn't optimized for natural language search. It is optimized for keywords, though. And you know what's great at generating keywords? LLMs!
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+Therefore, I decided to use a RAG pipeline to combine PubMed data with a LLM.
+
+## The PubMed RAG Pipeline
+
+Retrieval augmented generation, or RAG, is a way of giving a LLM context so it can better answer questions. 
+
+You pass the LLM some documents, along with a query, and prompt the LLM to use the documents when answering the question.
+
+![](RAG.png)
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:39]: Consider adding "alt text" to all images. If you'd like, you can add captions as well. Here's more information about the syntax: https://github.com/deepset-ai/haystack-home?tab=readme-ov-file#adding-an-image-in-article-content. 
```
![alt text here](image_path.png "Caption here")
```
You can change the image size with `#small`, `#medium` params as well

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+![](HealthcareChatbotArchitecture.png)
+
+```python
+from haystack import Pipeline
+from haystack.components.builders.prompt_builder import PromptBuilder
+
+keyword_prompt_template = """
+Your task is to convert the follwing question into 3 keywords that can be used to find relevant medical research papers on PubMed.
+Here is an examples:
+question: "What are the latest treatments for major depre###ssive disorder?"
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:96]: Was this intentional?
```suggestion
question: "What are the latest treatments for major depressive disorder?"
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+The management of PACS is challenging due to its heterogeneous presentation and the lack of evidence-based treatments. Current treatment approaches are mainly supportive and aim to alleviate symptoms and improve quality of life. These include:
+- Pulmonary rehabilitation for respiratory symptoms
+- Cardiac rehabilitation for cardiovascular symptoms
+- Cognitive-behavioral therapy for neurological symptoms
+- Dietary modifications and medications for gastrointestinal symptoms
+- Vaccination to prevent reinfection and further complications
+- Symptomatic treatment with medications such as nonsteroidal anti-inflammatory drugs (NSAIDs), corticosteroids, and antihistamines
+- Experimental treatments such as antiviral therapy, immunomodulatory therapy, and cell-based therapy
+```
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:152]: I think you can clarify this section by providing the code snippet to call the `pipeline.run` with the prompt and then providing the response. Something like this:

```python
question="What are the most current treatments for long COVID?"
result = pipeline.run({"prompt_builder":{"question": question}, .....})
print(result[...)
```
Output:
```shell
  The COVID-19 pandemic has led to a significant increase in the number of patients with post-acute COVID-19 syndrome (PACS), also known as long COVID. PACS is a complex, multisystem disorder that can affect various organs and systems, including the respiratory, cardiovascular, neurological, and gastrointestinal systems. The pathophysiology of PACS is not yet fully understood, but it is believed to be related to immune dysregulation, persistent inflammation, and microvascular injury.
The management of PACS is challenging due to its heterogeneous presentation and the lack of evidence-based treatments. Current treatment approaches are mainly supportive and aim to alleviate symptoms and improve quality of life. These include:
- Pulmonary rehabilitation for respiratory symptoms
- Cardiac rehabilitation for cardiovascular symptoms
- Cognitive-behavioral therapy for neurological symptoms
- Dietary modifications and medications for gastrointestinal symptoms
- Vaccination to prevent reinfection and further complications
- Symptomatic treatment with medications such as nonsteroidal anti-inflammatory drugs (NSAIDs), corticosteroids, and antihistamines
- Experimental treatments such as antiviral therapy, immunomodulatory therapy, and cell-based therapy
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+{% for article in articles %}
+  {{article.content}}
+  keywords: {{article.meta['keywords']}}
+  title: {{article.meta['title']}}
+{% endfor %}
+
+"""
+keyword_prompt_builder = PromptBuilder(template=keyword_prompt_template)
+prompt_builder = PromptBuilder(template=prompt_template)
+fetcher = PubMedFetcher()
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:121]: I believe the code snippets to initalize the Generators are missing, can you provide codes for them somewhere in the article?

**Review by bilgeyucel — COMMENTED** [review]
Great article @annthurium! 🚀 Your use case is super interesting and the article is right to the point. I left some minor comments 👍

**Comment by annthurium** [comment]
@bilgeyucel @TuanaCelik I believe I've addressed the feedback. Thanks again! 

My post still isn't showing up in the deploy preview, which makes me worried I'm missing some other mistakes. Is it because the date on the post is in the future?

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+Try it for yourself and see!
+
+```python
+pipe.run(data={"keyword_prompt_builder":{"question":question},
+               "prompt_builder":{"question": question},
+               "llm":{"generation_kwargs": {"max_new_tokens": 500}}})
+```
+
+```python
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:159]:
```suggestion
```shell
```

File: content/blog/mixtral-8x7b-healthcare-chatbot/index.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+What are the most current treatments for long COVID?
```
Comment (bilgeyucel) [inline content/blog/mixtral-8x7b-healthcare-chatbot/index.md:160]: Is this sentence the question? If so, it should be in the code snippet above.

```python
question="What are the most current treatments for long COVID?"
pipe.run(data={"keyword_prompt_builder":{"question":question},
               "prompt_builder":{"question": question},
               "llm":{"generation_kwargs": {"max_new_tokens": 500}}})
```

**Review by bilgeyucel — APPROVED** [review]
I left two small comments, apart from those, good to go!
